internal/policy/verify.go

```diff
@@ -97,43 +98,118 @@ func VerifyRelativeForRef(ctx context.Context, repo *git.Repository, initialPoli
-		// signatures are commit signatures. What do we do when metadata is
-		// signed using other methods? Do we instead flip the script and require
-		// metadata signatures to match git signing methods?
-		// UPDATE: This has likely been fixed. Policy RSL entries aren't
-		// verified. Instead, each policy state is internally verified AND each
-		// new policy state's root role signatures are verified with the prior
-		// policy's root role to ensure they have a threshold of valid
-		// signatures. This comment is being left in until this workflow is
-		// audited against the gittuf specification.
-		if entry.RefName == PolicyRef {
-			// TODO: this is repetition if the firstEntry is for policy
-			newPolicy, err := LoadStateForEntry(ctx, repo, entry)
+	// 3. Verify each entry, looking for a fix when an invalid entry is encountered
+	var invalidEntry *rsl.ReferenceEntry
+	var verificationErr error
+	for len(entries) != 0 {
+		if invalidEntry == nil {
+			// Pop entry from queue
+			entry := entries[0]
+			entries = entries[1:]
+
+			if entry.RefName == PolicyRef {
+				// TODO: this is repetition if the firstEntry is for policy
+				newPolicy, err := LoadStateForEntry(ctx, repo, entry)
+				if err != nil {
+					return err
+				}
+
+				if err := currentPolicy.VerifyNewState(ctx, newPolicy); err != nil {
+					return err
+				}
+
+				currentPolicy = newPolicy
+				continue
+			}
+
+			if err := verifyEntry(ctx, repo, currentPolicy, entry); err != nil {
+				// If the invalid entry is never marked as skipped, we return err
+				if !entry.SkippedBy(annotations[entry.ID]) {
+					return err
+				}
+
+				// The invalid entry's been marked as skipped but we still need
+				// to see if another entry fixed state for non-gittuf users
+				invalidEntry = entry
+				verificationErr = err
+			}
+			continue
+		}
+
+		// This is only reached when we have an invalid state
+
+		// 1. What's the last good state?
+		// We don't care about its annotations because it's been verified prior
+		// to verifying the current entry, we also know it's unskipped
+		lastGoodEntry, _, err := rsl.GetLatestUnskippedReferenceEntryForRefBefore(repo, invalidEntry.RefName, invalidEntry.ID)
+		if err != nil {
+			return err
+		}
+		lastGoodEntryCommit, err := repo.CommitObject(lastGoodEntry.TargetID)
+		if err != nil {
+			return err
+		}
+		lastGoodTreeID := lastGoodEntryCommit.TreeHash
+
+		// 2. What entries do we have in the current verification set for the
+		// ref? The first one that is tree-same as lastGoodEntry's commit is the
+		// fix. Entries prior to that one in the queue are considered invalid
+		// and must be skipped
+		fixed := false
+		unskippedInvalid := false
+		newEntryQueue := []*rsl.ReferenceEntry{}
+		for len(entries) != 0 {
+			newEntry := entries[0]
+			entries = entries[1:]
+
+			// TODO: this is essentially just refName or policy entries, right?
+			if newEntry.RefName != invalidEntry.RefName {
```

Comment:
The full list of entries is provided by `rsl.GetReferenceEntriesInRangeForRef`. It returns ref entries for the requested ref and any entries for the gittuf namespace. I think preserving entries that aren't for the affected ref to be processed separately is fine in general because we may one day open this up to verifying multiple refs and not just one?